server/src/main/java/timely/store/iterators/TimeSeriesGroupingIterator.java

```diff
@@ -0,0 +1,309 @@
+            result += (filters[i] * e.getValue());
+            i++;
+        }
+        LOG.trace("compute - result: {}", result);
+        return result;
+    }
+
+    @Override
+    public Key getTopKey() {
+        return topKey;
+    }
+
+    @Override
+    public Value getTopValue() {
+        return topValue;
+    }
+
+    @Override
+    public boolean hasTop() {
+        LOG.trace("hasTop()");
+        if (seriesIterator.hasNext()) {
+            LOG.trace("seriesIterator - hasNext()");
+            Pair<Key, Double> p = seriesIterator.next();
+            topKey = p.getFirst();
+            topValue = new Value(MetricAdapter.encodeValue(p.getSecond()));
+            return true;
+        } else {
+            try {
+                refillBuffer();
+                if (series.size() == 0) {
+                    return false;
+                } else {
+                    seriesIterator = series.iterator();
+                    Pair<Key, Double> p = seriesIterator.next();
+                    topKey = p.getFirst();
+                    topValue = new Value(MetricAdapter.encodeValue(p.getSecond()));
+                    return true;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException("Error filling buffer", e);
+            }
+        }
+    }
+
+    @Override
+    public void next() throws IOException {
+        LOG.trace("next()");
+    }
+
+    @Override
+    public void init(SortedKeyValueIterator<Key, Value> source, Map<String, String> options, IteratorEnvironment env)
+            throws IOException {
+        super.init(source, options, env);
+        String filterOption = options.getOrDefault(FILTER, null);
+        if (null == filterOption) {
+            throw new IllegalArgumentException("Window size must be specified.");
+        }
+        String[] split = filterOption.split(",");
+        filters = new Double[split.length];
+        for (int i = 0; i < split.length; i++) {
+            filters[i] = Double.parseDouble(split[i]);
+        }
+        LOG.trace("init - filter: {}", Arrays.toString(filters));
+    }
+
+    @Override
+    public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
+        super.seek(range, columnFamilies, inclusive);
+        series.clear();
+        fillBuffer();
+        seriesIterator = series.iterator();
+    }
+
+    private void refillBuffer() throws IOException {
+        LOG.trace("refill()");
+        Metric top = null;
+        while (super.hasTop()) {
+            Key k = super.getTopKey();
+            Metric m = MetricAdapter.parse(k, super.getTopValue());
+            timely.model.Value v = m.getValue();
+            m.setValue(null);
+            Collections.sort(m.getTags());
+            if (top != null && top.equals(m)) {
+                // we have cycled back to the first series
+                break;
```

Comment:
Stop the loop when you receive a metric that is less than or equal to the previous metric. Change made.